adc/ade7913/ade7913-driver.c

```diff
@@ -605,131 +832,145 @@ static void msg_loop(void)
-			log_error("failed to initialize edma");
-			resourceDestroy(common.gpio_irq_lock);
-			resourceDestroy(common.gpio_irq_cond);
-
+		res = pwm_init();
+		if (res < 0) {
+			log_error("Failed to initialize PWM");
 			break;
 		}
 
-		if ((res = edma_configure()) != 0) {
-			resourceDestroy(common.gpio_irq_lock);
-			resourceDestroy(common.gpio_irq_cond);
+		res = edma_init(edma_error_handler);
+		if (res < 0) {
+			log_error("Failed to initialize eDMA");
+			break;
+		}
 
-			log_error("failed to configure edma");
+		res = edma_configure();
+		if (res < 0) {
+			log_error("Failed to configure eDMA");
 			break;
 		}
+	} while (0);
 
-		if ((res = dev_init()) < 0) {
-			log_error("device initialization failed");
-			resourceDestroy(common.gpio_irq_lock);
-			resourceDestroy(common.gpio_irq_cond);
-			resourceDestroy(common.edma_spi_rcv_lock);
+	if (res < 0) {
+		if (common.oid.port != (uint32_t)-1) {
+			portDestroy(common.oid.port);
+		}
+		if (common.edma_spi_rcv_lock != (handle_t)-1) {
 			resourceDestroy(common.edma_spi_rcv_lock);
+		}
+		if (common.edma_spi_rcv_cond != (handle_t)-1) {
+			resourceDestroy(common.edma_spi_rcv_cond);
+		}
+		if (common.dready_cond != (handle_t)-1) {
 			resourceDestroy(common.dready_cond);
+		}
+		if (common.buff != MAP_FAILED) {
 			munmap(common.buff, (ADC_BUFFER_SIZE + _PAGE_SIZE - 1) / _PAGE_SIZE * _PAGE_SIZE);
-
-			break;
 		}
-
-		/* Enable SPI edma receive request */
-		*(common.spi1_ptr + spi_der) = (1 << 1);
-		/* Enable GPIO DREADY interrupt */
-		*(common.gpio3_ptr + gpio_imr) |= (0b1 << 20);
-	} while (0);
+	}
 
 	return res;
 }
 
 
 int main(int argc, char **argv)
 {
-	int i;
-	oid_t root;
+	oid_t tmp_oid;
+	int i, err;
+
+	priority(ADE7913_PRIO);
 
-	common.gpio3_ptr = (void *)0x40134000;
-	common.spi1_ptr = (void *)0x40114000;
+	memset(&common, 0, sizeof(common));
 
 	if (argc != 3) {
 		log_error("No device list or no SPI device given");
-		return -EINVAL;
+		return EXIT_FAILURE;
 	}
-	else {
-		common.order = argv[1];
-		common.devcnt = strlen(argv[1]);
-		common.spi = atoi(argv[2]);
-
-		for (i = 0; i < common.devcnt; ++i) {
-			if ((int)(common.order[i] - '0') >= common.devcnt || (int)(common.order[i] - '0') < 0) {
-				log_error("Wrong order format provided");
-				return -1;
-			}
-		}
 
-		if (common.spi < 0) {
-			log_error("Wrong spi number provided");
-			return -1;
-		}
+	common.spi = atoi(argv[2]);
+	if (common.spi < 1 || common.spi > sizeof(spi_base) / sizeof(spi_base[0])) {
+		log_error("Wrong spi number provided");
+		return EXIT_FAILURE;
+	}
 
-		log_info("Device order: %s", common.order);
+	common.oid.port = (uint32_t)-1;
+	common.edma_spi_rcv_lock = (handle_t)-1;
+	common.edma_spi_rcv_cond = (handle_t)-1;
+	common.dready_cond = (handle_t)-1;
+	common.buff = MAP_FAILED;
+
+	common.spi_ptr = (uint32_t *)spi_base[common.spi - 1];
+	common.gpio3_ptr = (uint32_t *)0x40134000;
+	common.iomux_ptr[0] = (uint32_t *)0x400e8000 + 4 + pctl_mux_gpio_ad_18;
```

Comment:
Why is this `+ 4` needed? Also, you cast to `uint32_t *` first, so the result is (uint32_t *)0x400e8000 + 4 = 0x400e8010, is this intended? If so, this might be somewhat misleading.

Reply:
Yes it is correct. `IOMUXC` base addres is `0x400e8000`, while the first "documented" GPIO MUX control register is for PAD_GPIO_EMC_B1_00 and it is located at offset `4 * sizeof(uint32_t)`, therefore like `pctl_mux_gpio_ad_18 = 81` its MUX address is `0x400e8154`.
Btw. [phoenix-rtos-kernel/hal/armv7m/imxrt/117x/imxrt117x.c#L96](https://github.com/phoenix-rtos/phoenix-rtos-kernel/blob/master/hal/armv7m/imxrt/117x/imxrt117x.c#L96)

Reply:
Touché